lib/saluki-core/src/runtime/process.rs

```diff
@@ -0,0 +1,276 @@
+use std::{
+    future::Future,
+    ops::Deref,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering::Relaxed},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+
+use memory_accounting::allocator::{AllocationGroupRegistry, AllocationGroupToken, Track as _, Tracked};
+use pin_project::pin_project;
+use tracing::{debug_span, instrument::Instrumented, Instrument as _};
+
+static GLOBAL_PROCESS_ID_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+/// Process identifier.
+///
+/// A simple, numeric identifier that uniquely identifies a process.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct Id(usize);
+
+impl Id {
+    /// Creates a new process identifier.
+    pub fn new() -> Self {
+        let id = GLOBAL_PROCESS_ID_COUNTER.fetch_add(1, Relaxed);
+        Self(id)
+    }
+
+    fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+/// Process name.
+///
+/// A human-readable name for a process that only contains alphanumeric characters, underscores, and periods.
+///
+/// Process names are scoped, such that the resulting process name is nested. For example, if a supervisor has a name of
+/// `topology_sup`, and a child process is added to that supervisor with a name of `worker`, the resulting process name
+/// for the child process will be `topology_sup.worker`. Process names can be arbitrarily nested in this way.
+///
+/// Process names will be sanitized if they contain invalid characters, such as hyphens or spaces. Invalid characters
+/// will be replaced with underscores.
```

Comment:
I'm not opposed, just curious why we don't reject invalid names outright? Haven't confirmed yet, it's possible the intention here is for process names to arrive via user config, in which case this does make some sense.

Reply:
Yeah, more or less. Process names will come, in part, from user-defined identifiers that might be provided at a layer above (or higher) the point where processes themselves are constructed, so we want to be more liberal in what we accept so that code doesn't have to bubble up that sort of constraint.